docs/advanced-topics/custom-vertex-id.md

```diff
@@ -0,0 +1,92 @@
+The second option allows to you either provide a long-type ID or a string-type ID. Note that
+if you turn on `graph.allow-string-vid`, you must also turn on `graph.set-vertex-id`.
+
+Both of the two options are `GLOBAL_OFFLINE`, with default values being `false`. Being
+global offline means once they are set, you would have to bring down the entire JanusGraph
+cluster to alter the value. See the below two scenarios.
+
+### Create a new Graph
+
+Simply set `graph.set-vertex-id=<true/false>` and `graph.allow-string-vid=<true/false>` in your config file.
+When the graph is created, this value is loaded and persisted. After the graph is created,
+changing the option value in the config file would make no difference as JanusGraph simply ignores the value.
+
+### Alter an existing Graph
+
+To alter a global offline config (not limited to the aforementioned two options), you need to shut
+down all JanusGraph instances. Then open a gremlin console, connect to JanusGraph instance and run
+the following:
+
+```groovy
+mgmt = graph.openManagement();
+mgmt.set("graph.set-vertex-id", true);
+// optional, if you want to provide string ID
+mgmt.set("graph.allow-string-vid", true);
+mgmt.commit();
+```
+
+It will return an error if you have open JanusGraph instances. Otherwise, it will succeed and
+now your global configs would have been updated.
+
+## Custom Long ID
+
+To provide a long vertex ID, you must call `graph.getIDManager().toVertexId(long)` to
+retrieve a transformed JanusGraph ID. The input number must be positive. This is because
+certain range of long IDs are reserved for internal usage, so you have to always call
+the aforementioned utility to convert your ID. Unfortunately, this is not supported in
+all gremlin clients as it is a JanusGraph specific utility. You can create vertices like
+below:
+
+```groovy
+g = graph.traversal()
+g.addV().property(T.id, graph.getIDManager().toVertexID(123L)).next()
+g.tx().commit()
+```
+
+Note that if you want to get back the original ID you provided, you need to call
+`graph.getIDManager().fromVertexID(long)`.
+
+## Custom String ID
+
+A string vertex ID can have arbitrary length as long as the underlying storage backend
+permits. The string value must consist of ASCII characters only, and it cannot contain
+JanusGraph reserved relation delimiter. By default, this reserved character is `-` (dash),
+which means you cannot use UUID string as a vertex ID since it contains a `-` character.
+However, JanusGraph allows you to set `JANUSGRAPH_RELATION_DELIMITER` system property
+which can be any visible ASCII character. Once it is set, JanusGraph will prohibit that
+specific character instead of the default `-` character. Alternatively, you can set
+`JANUSGRAPH_RELATION_DELIMITER` environment variable, which is evaluated if and only if
+the `JANUSGRAPH_RELATION_DELIMITER` system property is not present. For example, you can
+use the following command to set system property when opening gremlin console:
```

Comment:
> but if user uses non-printable characters for ID and printable character for delimiter wouldn't it wrongly printer in the Gremlin Console anyway

Good point. In the beginning, I thought we should allow all ASCII characters but later I changed my mind, leaving the requirements inconsistent. I am thinking of requiring readable ASCII characters only.

Reply:
I’m just thinking if this requirement is only for better visibility in Gremlin console then it could be we don’t need this restriction actually. TinkerPop supports Object as id for vertices, meaning that it can be a String with non-printable characters. The fact they will be harder to copy paste manually via Gremlin console isn’t a deal breaker I think. Users need to understand that if they use non-printable characters - they logically won’t see those characters, but it shouldn’t stop them from using those characters because they can still be correctly used inside JanusGraph. As far as I understand it doesn’t matter for JanusGraph if key characters or relation delimiter is printable or non-printable. The only requirement for the delimiter is to be different from the rest of the characters.
I might be wrong though. Just brainstorming.

Reply:
I'm now thinking, as this actually not a breaking change on the storage layer, we could really allow using only printable ASCII characters for Strings as for now, but if users claim they need to use non-printable ASCII characters we can always relax the checks and allow any ASCII characters. 
This way we will not deal with potential problems printing those ids on Gremlin Console side, and have room to relax id requirements later if users want to.

Reply:
My arguments:

1) Users might accidentally introduce non-printable ASCII characters and this would make debugging difficult. For example, they might accidentally enter a new line on the gremlin console when they enter the vertex ID and still succeed.
2) We could easily extend support for printable ASCII characters to all ASCII characters, without introducing a breaking change.
3) Possible implications on indexing backends. Some secondary indexing backends might not allow non-printable ASCII characters (I didn't verify, so this might not actually hold).

Reply:
Agree with your points here. Let's use only printable characters as a safety measure as for now (both for delimiter and for actual id). Later on if we receive feedback that anyone needs to use non-printable characters for some use-cases then we could introduce a special feature flag which relaxes String requirements and allows any characters to be used.